spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaAutoConfiguration.java

```diff
@@ -137,4 +146,48 @@ public KafkaAdmin kafkaAdmin() {
 		return kafkaAdmin;
 	}
 
+	@Bean
+	@ConditionalOnProperty(name = "spring.kafka.retry.topic.enabled")
+	@ConditionalOnMissingBean
+	public RetryTopicConfiguration kafkaRetryTopicConfiguration(KafkaOperations<Object, Object> kafkaOperations) {
+		RetryTopicConfigurationBuilder builder = RetryTopicConfigurationBuilder.newInstance();
+		KafkaProperties.Retry.Topic retryTopic = this.properties.getRetry().getTopic();
+		PropertyMapper map = PropertyMapper.get().alwaysApplyingWhenNonNull();
+		map.from(retryTopic.getAttempts()).to(builder::maxAttempts);
+		map.from(createBackOffFor(retryTopic)).to(builder::customBackoff);
+		return builder.create(kafkaOperations);
+	}
+
+	private SleepingBackOffPolicy<?> createBackOffFor(KafkaProperties.Retry.Topic properties) {
+		Long min = properties.getDelayMillis();
+		Long max = properties.getMaxDelayMillis();
+		Double multiplier = properties.getMultiplier();
+		Boolean isRandom = properties.isRandomBackOff();
+		if (min == null && max == null && multiplier == null && isRandom == null) {
```

Comment:
This doesn't look quite right to me. If random is non-null, it'll cause a `FixedBackOffPolicy` to be returned but the random setting will never be used. I think the `isRandom == null` check could just be removed here. WDYT, @tomazfernandes?

Reply:
Sure, we can do that, thanks for pointing it out. In the end, the result will also be a `FixedBackOffPolicy`, but the way you're proposing makes the behavior clearer -  that we're in fact ignoring the isRandom flag.

Actually, I'd like to ask you a question about this, if that's ok. This logic was mainly brought from the [Spring Retry](https://github.com/spring-projects/spring-retry/blob/302f7999932421d39800b81b5fa9f6b1c557b957/src/main/java/org/springframework/retry/annotation/AnnotationAwareRetryOperationsInterceptor.java#L404) project, but it's private there as you can see in the link, so we have a slightly modified copy in the [Spring Kafka](https://github.com/spring-projects/spring-kafka/blob/43cb78127b3851c03037418533bf393c2fc7db3a/spring-kafka/src/main/java/org/springframework/kafka/annotation/RetryableTopicAnnotationProcessor.java#L171
) project, but it's not accessible there either, that's why I made another copy here.

I'm wondering if maybe it'd be better to open up this logic in Spring Kafka, and use it here, so we can keep the exact same behavior whether the user is configuring through auto configuration or through the `@RetryableTopic` annotation. I've been meaning to review a few things there too - maybe let the user set a `delay` of 0 and get a `NoBackOffPolicy` for example, besides a couple of issues I've noticed recently.

That might hold this feature back a while though, until SK with this change is released, and Boot is updated to build against that version. And since we won't make any breaking changes there too, maybe it's something we can change in a later version.

WDYT?

Thanks

Reply:
I think that's a great idea as it'll avoid any drift between the two projects with how the four settings are mapping to a policy. Let's wait for that API to appear in Spring Kafka and then we can update the proposal here to use it.

Reply:
Ok, sounds good, I’ll open an issue for this change there. Thanks @wilkinsona.

Reply:
Hi @wilkinsona, we addressed this issue [here](https://github.com/spring-projects/spring-retry/pull/277) in the `Spring Retry` project.

Not sure how the release schedule is there now, but I'll update this PR when the API appears.

Thanks.

Reply:
@tomazfernandes The changes in Spring Retry should now be available in 1.3.3-SNAPSHOT. Do you have time to have another look at this before next week's Spring Boot 2.7.0-RC1 release?

Reply:
Hi @wilkinsona, that's good news, sure, thanks for letting me know. I should have these changes soon.

Reply:
@wilkinsona it seems [build.gradle](https://github.com/spring-projects/spring-boot/blob/3485070aeef0b5ebff0cf4dfcc76381a82cf4f6a/spring-boot-project/spring-boot-dependencies/build.gradle#L1421-L1427) is still pointing to `1.3.2`.

Is it going to be changed to `1.3.3-SNAPSHOT`, or should I do that as part of this PR?

Thanks

Reply:
Please change it in the PR for now. We can straighten things out as part of the merge.

Reply:
@tomazfernandes this PR targets the `2.7.x` branch and we've switched to snapshot there https://github.com/spring-projects/spring-boot/blob/f96efa72abbc8ab4abd6f5378ebc9574a84c30b1/spring-boot-project/spring-boot-dependencies/build.gradle#L1783-L1787

Reply:
Also, I don't seem to be able to cleanly either rebase my branch onto origin/2.7.x., or merge that branch into mine. Do you have any pointers on that? Thanks again

Reply:
We prefer forward merges to back-ports, however, if switching the base branch to 2.7.x is proving tricky, please leave it as main and switch to Spring Retry 1.3.3-SNAPSHOT in your PR. We can then take care of getting the changes based on top of 2.7.x as part of merging the contribution.

Reply:
I've submitted the changes. I think this looks a lot better overall than the previous code, thanks mainly to the great guidance I've been receiving. Please let me know if there's anything you'd like me to change.

Besides using the `BackOffPolicyBuilder`, I've also changed a few things in the `sensible defaults`, more on par with what I've came to understand about the auto configuration role.

Now we have:

`useSingleTopicForFixedDelays()`
`suffixTopicsWithIndexValues()`
`doNotAutoCreateRetryTopics()`

With these settings, by default users will get 3 topics, independently of the number of `attempts` or `delay` value:
`myTopic`, `myTopic-retry` and `myTopic-dlt`

If a `multiplier` or `maxDelay` is set, yelding multiple delay values, users will get one topic per attempt plus the dlt:
`myTopic`, `myTopic-retry-0`, `myTopic-retry-1`, `myTopic-retry-2` ... `myTopic-dlt`

A Kafka topic can only handle a single delay value due to message ordering - if delays were different within a topic, we couldn't tell if a record after the one being processed was supposed to be processed earlier.

Note that although we're setting the feature not to create the topics, Kafka's `Consumer` can still auto create them if configured to do so.

@garyrussell and @artembilan, if possible, please let me know if you agree with these defaults and with the code changes.

We can add more features to the auto configuration as users request them, but if there's anything that comes to mind that should be in this first version I can add them.

Thanks, looking forward to seeing this in place!

Reply:
@wilkinsona, there's a conflict on the `build.gradle` file regarding the `spring-retry` version. Should I try to resolve the conflict here on GitHub, or is it something you should do? It would be interesting so we can see the test results.

Thanks

Reply:
Also, sorry for my not-so-great git skills - this is a complex project and I couldn't solve this in a proper way with the solutions I know. I appreciate the assistance.

Reply:
I solved the git issue by checking out the 2.7.x brach, cherry-picking my commits, and hard reseting this branch to that. No clue on why I couldn't do a clean merge or rebase though - have you rebased this branch at some point maybe?

Also, I don't seem to be able to do a clean build locally, there's always a failing test. I've tried different JDK versions / implementations, cleaned up my Gradle caches, and though the failing tests change, the build never gets through cleanly. Maybe my computer is not powerful enough to handle such intensive testing. If you have any pointers on that it'd be great, otherwise I'll keep trying.

The tests in the `kafka` package are ok when running locally though, so it should be ok here.

Thanks!

Reply:
> @tomazfernandes this PR targets the `2.7.x` branch and we've switched to snapshot there
> 
> https://github.com/spring-projects/spring-boot/blob/f96efa72abbc8ab4abd6f5378ebc9574a84c30b1/spring-boot-project/spring-boot-dependencies/build.gradle#L1783-L1787

I've changed the PR to target 2.7.x. - hope it's ok now.

Thanks